utils/rollout/rolloututil.go

```diff
@@ -39,6 +45,19 @@ func isGenerationObserved(ro *v1alpha1.Rollout) bool {
 	return int64(observedGen) == ro.Generation
 }
 
+func isWorkloadGenerationObserved(ro *v1alpha1.Rollout) bool {
+	if _, ok := annotations.GetWorkloadGenerationAnnotation(ro); !ok {
+		return true
+	}
+	workloadGeneration, _ := annotations.GetWorkloadGenerationAnnotation(ro)
+	observedWorkloadGen, err := strconv.Atoi(ro.Status.WorkloadObservedGeneration)
+	if err != nil {
+		return true
+	}
+
+	return int32(observedWorkloadGen) == workloadGeneration
```

Comment:
1. The code scanner is catching a valid concern with the use of strconv.Atoi. We should use `strconv.ParseInt(ro.Status.WorkloadObservedGeneration, 10, 32)` instead

2. I just noticed that we may not handle the case properly where we still may have a workloadGenerationAnnotation on the rollout (leftover from when it switched from workload to non-workload ref) even if rollout switched to using inline template.

I think this function needs to check if `ro.Spec.WorkloadRef != nil` before proceeding. At the same time, I think we should remove the workload generation annotation from the rollout, since I don't think that gets cleaned up at the moment.

Reply:
Thanks for the feedback. The PR is updated and an e2e test case is added to verify the workload generation annotation  is removed after it switches to inline template.